sql/analyzer/pushdown.go

```diff
@@ -801,139 +803,108 @@ func getIndexesByTable(ctx *sql.Context, a *Analyzer, node sql.Node, scope *Scop
 			}
-		}
-		if pkIndex == nil {
-			return s, transform.SameTree, nil
-		}
-
-		// Get primary key column names; these are qualified
-		pkColNames := pkIndex.Expressions()
-
-		// Extract SortField Column Names
-		var sfColNames []string
-		for _, field := range s.SortFields {
-			gf, ok := field.Column.(*expression.GetField)
+			table := rs.Table
+			if w, ok := table.(sql.TableWrapper); ok {
+				table = w.Underlying()
+			}
+			idxTbl, ok := table.(sql.IndexAddressableTable)
 			if !ok {
 				return n, transform.SameTree, nil
 			}
-			// Resolve aliases; aliases should have empty table in GetField
-			if name, ok := aliasMap[gf.String()]; ok {
-				sfColNames = append(sfColNames, name)
-			} else {
-				sfColNames = append(sfColNames, gf.String())
+			idxs, err := idxTbl.GetIndexes(ctx)
+			if err != nil {
+				return nil, transform.SameTree, err
 			}
-		}
 
-		// SortField is definitely not a prefix to PrimaryKey
-		if len(sfColNames) > len(pkColNames) {
-			return s, transform.SameTree, nil
-		}
-
-		// Check if SortField is a prefix to PrimaryKey
-		for i := 0; i < len(sfColNames); i++ {
-			// Stop when column names stop matching
-			if sfColNames[i] != pkColNames[i] {
-				return s, transform.SameTree, nil
+			var pkIndex sql.Index // TODO: support secondary indexes
+			for _, idx := range idxs {
+				if idx.ID() == "PRIMARY" {
+					pkIndex = idx
+					break
+				}
+			}
+			if pkIndex == nil {
+				return n, transform.SameTree, nil
+			}
+			// Some Primary Keys (like doltHistoryTable) are not in order
+			if oi, ok := pkIndex.(sql.OrderedIndex); ok && (!oi.Reversible() || oi.Order() == sql.IndexOrderNone) {
+				return n, transform.SameTree, nil
 			}
-		}
 
-		// Create lookup based off of PrimaryKey
-		indexBuilder := sql.NewIndexBuilder(pkIndex)
-		lookup, err := indexBuilder.Build(ctx)
-		if err != nil {
-			return nil, transform.SameTree, err
-		}
-		if !pkIndex.CanSupport(lookup.Ranges...) {
-			return n, transform.SameTree, nil
-		}
-		newNode, err := plan.NewStaticIndexedAccessForResolvedTable(rs, lookup)
-		if err != nil {
-			return nil, transform.SameTree, err
-		}
+			pkColNames := pkIndex.Expressions()
+			if len(sfExprs) > len(pkColNames) {
+				return n, transform.SameTree, nil
+			}
+			for i, fieldExpr := range sfExprs {
+				if s.SortFields[0].Order != s.SortFields[i].Order {
+					return n, transform.SameTree, nil
+				}
+				fieldName := fieldExpr.String()
+				if alias, ok := sfAliases[strings.ToLower(pkColNames[i])]; ok && alias == fieldName {
+					continue
+				}
+				if pkColNames[i] != fieldExpr.String() {
+					return n, transform.SameTree, nil
+				}
+			}
 
-		var resNode sql.Node = newNode
-		if decoratingParent != nil {
-			resNode, err = decoratingParent.WithChildren(resNode)
+			// Create lookup based off of PrimaryKey
+			indexBuilder := sql.NewIndexBuilder(pkIndex)
+			lookup, err := indexBuilder.Build(ctx)
+			lookup.IsReverse = s.SortFields[0].Order == sql.Descending
```

Comment:
only first sort field? some of this looks kind of sketch but hard to verify on inspection. Second round of testing edge cases would help